File: src/0931_minimum_falling_path_sum.js
class ListNode {
  constructor(val = 0, next = null) { this.val = val; this.next = next; }
}

class TreeNode {
  constructor(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
}

class Node {
  constructor(val = 0, next = null, random = null, left = null, right = null, neighbors = null, key = -1) {
    this.val = val;
    this.next = next;
    this.random = random;
    this.left = left;
    this.right = right;
    this.neighbors = neighbors === null ? [] : neighbors;
    this.key = key;
  }
}

function listNodeToString(node) {
  const parts = [];
  for (let cur = node; cur !== null; cur = cur.next) parts.push(String(cur.val));
  return parts.length ? parts.join('->') : 'None';
}

function treeToString(node) {
  if (node === null || node === undefined) return 'None';
  return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
}

function buildListNode(values) {
  const dummy = new ListNode();
  let tail = dummy;
  for (const value of values) {
    tail.next = new ListNode(value);
    tail = tail.next;
  }
  return dummy.next;
}

class MinHeap {
  constructor(compare = (a, b) => a < b) { this.data = []; this.compare = compare; }
  get length() { return this.data.length; }
  peek() { return this.data[0]; }
  push(value) {
    this.data.push(value);
    this._up(this.data.length - 1);
  }
  pop() {
    if (this.data.length === 0) return undefined;
    const root = this.data[0];
    const last = this.data.pop();
    if (this.data.length > 0) {
      this.data[0] = last;
      this._down(0);
    }
    return root;
  }
  _up(i) {
    while (i > 0) {
      const p = Math.floor((i - 1) / 2);
      if (!this.compare(this.data[i], this.data[p])) break;
      [this.data[i], this.data[p]] = [this.data[p], this.data[i]];
      i = p;
    }
  }
  _down(i) {
    while (true) {
      const l = i * 2 + 1;
      const r = l + 1;
      let best = i;
      if (l < this.data.length && this.compare(this.data[l], this.data[best])) best = l;
      if (r < this.data.length && this.compare(this.data[r], this.data[best])) best = r;
      if (best === i) break;
      [this.data[i], this.data[best]] = [this.data[best], this.data[i]];
      i = best;
    }
  }
}

class UnionFind {
  constructor(n = null) {
    this.f = new Map();
    this.par = n === null ? [] : Array.from({ length: n }, (_, i) => i);
    this.rank = n === null ? [] : Array(n).fill(1);
    this.size = n === null ? [] : Array(n).fill(1);
    this.count = n === null ? 0 : n;
  }
  findParent(x) {
    const y = this.f.has(x) ? this.f.get(x) : x;
    if (x !== y) {
      const root = this.findParent(y);
      this.f.set(x, root);
      return root;
    }
    return y;
  }
  find(x) {
    if (this.par.length === 0) return this.findParent(x);
    while (x !== this.par[x]) {
      this.par[x] = this.par[this.par[x]];
      x = this.par[x];
    }
    return x;
  }
  union(x, y) {
    if (this.par.length === 0) {
      const px = this.findParent(x), py = this.findParent(y);
      if (px === py) return false;
      this.f.set(px, py);
      return true;
    }
    const px = this.find(x), py = this.find(y);
    if (px === py) return false;
    if (this.rank[px] > this.rank[py]) {
      this.par[py] = px;
      this.rank[px] += this.rank[py];
      this.size[px] += this.size[py];
    } else {
      this.par[px] = py;
      this.rank[py] += this.rank[px];
      this.size[py] += this.size[px];
    }
    this.count -= 1;
    return true;
  }
}

function minFallingPathSum(matrix) {
  const Memo = new Map();
  function Path(i, k, n) {
    const key = `${i},${k}`;
    if (Memo.has(key)) return Memo.get(key);
    if (i === n - 1) return matrix[i][k];
    if (k > 0 && k < n - 1) {
      const Psx = matrix[i][k] + Path(i + 1, k - 1, n);
      const Pst = matrix[i][k] + Path(i + 1, k, n);
      const Pdx = matrix[i][k] + Path(i + 1, k + 1, n);
      Memo.set(key, Math.min(Math.min(Pdx, Pst), Psx));
    } else if (k === 0) {
      const Pst = matrix[i][k] + Path(i + 1, k, n);
      const Pdx = matrix[i][k] + Path(i + 1, k + 1, n);
      Memo.set(key, Math.min(Pst, Pdx));
    } else {
      const Psx = matrix[i][k] + Path(i + 1, k - 1, n);
      const Pst = matrix[i][k] + Path(i + 1, k, n);
      Memo.set(key, Math.min(Pst, Psx));
    }
    return Memo.get(key);
  }
  let Min = 10 ** 7;
  for (let k = 0; k < matrix[0].length; k++) {
    const CP = Path(0, k, matrix[0].length);
    if (CP < Min) Min = CP;
  }
  return Min;
}

module.exports = { minFallingPathSum };

if (require.main === module) {
  const assert = require('assert');
  assert.strictEqual(minFallingPathSum([[2, 1, 3], [6, 5, 4], [7, 8, 9]]), 13);
  assert.strictEqual(minFallingPathSum([[-19, 57], [-40, -5]]), -59);
}

File: src/0931_minimum_falling_path_sum.rs
use std::collections::HashMap;

struct Solution;

impl Solution {
    pub fn min_falling_path_sum(matrix: Vec<Vec<i32>>) -> i32 {
        let mut memo = HashMap::<(usize, usize), i32>::new();
        let mut min_value = 10_000_000;
        for k in 0..matrix[0].len() {
            let current_path = Self::path(0, k, matrix[0].len(), &matrix, &mut memo);
            if current_path < min_value {
                min_value = current_path;
            }
        }
        min_value
    }

    fn path(
        i: usize,
        k: usize,
        n: usize,
        matrix: &[Vec<i32>],
        memo: &mut HashMap<(usize, usize), i32>,
    ) -> i32 {
        if let Some(value) = memo.get(&(i, k)) {
            return *value;
        }
        if i == n - 1 {
            return matrix[i][k];
        }

        let value = if k > 0 && k < n - 1 {
            let psx = matrix[i][k] + Self::path(i + 1, k - 1, n, matrix, memo);
            let pst = matrix[i][k] + Self::path(i + 1, k, n, matrix, memo);
            let pdx = matrix[i][k] + Self::path(i + 1, k + 1, n, matrix, memo);
            pdx.min(pst).min(psx)
        } else if k == 0 {
            let pst = matrix[i][k] + Self::path(i + 1, k, n, matrix, memo);
            let pdx = matrix[i][k] + Self::path(i + 1, k + 1, n, matrix, memo);
            pst.min(pdx)
        } else {
            let psx = matrix[i][k] + Self::path(i + 1, k - 1, n, matrix, memo);
            let pst = matrix[i][k] + Self::path(i + 1, k, n, matrix, memo);
            pst.min(psx)
        };

        memo.insert((i, k), value);
        value
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mirrors_python_main_assertions() {
        assert_eq!(
            Solution::min_falling_path_sum(vec![vec![2, 1, 3], vec![6, 5, 4], vec![7, 8, 9]]),
            13
        );
        assert_eq!(
            Solution::min_falling_path_sum(vec![vec![-19, 57], vec![-40, -5]]),
            -59
        );
    }
}
